### ISSUE TITLE
Subscribing to the Event System

### ISSUE THREAD

**Issue by kwizzn** [trufflesuite/truffle#4656]
- [ ] I've [opened a support ticket](https://trufflesuite.zendesk.com/hc/en-us/requests/new) before filing this issue.

---------------------------

## Issue

I know the Event System is flagged "experimental" but as I understand the documentation, experiments and feedback are welcome so here it comes: I have a use case where I need `truffle compile` to make minor adjustments to some files in `/contracts` before running the actual compilation. I cannot get `compile:start` handler to fire, though.

## Steps to Reproduce

In my `truffle-config.js`, I added the `subscribers` key:
```javascript
module.exports = {
    [...]
    subscribers: {
        testnetOverride: require('./testnet-override.js'),
    }
}
```

In the project root, I added `testnet-override.js`:
```javascript
module.exports = {
    handlers: {
        "compile:start": [
            function () {
                console.log('foobar');
            }
        ],
    },
};
```

## Expected Behavior

I expect to see the `foobar` log output before the contracts are compiled.

## Actual Results

I get the regular compilation output with no errors but also no log output.

## Environment

* Operating System: MacOS 12.1
* Ethereum client: truffle develop
* Truffle version (`truffle version`): v5.4.30
* node version (`node --version`): v16.13.0
* npm version (`npm --version`): v8.1.0

**Comment by gnidan** [comment]
Neat! Could you share your completed example? I'd love to see how you're rewriting your contracts mid-flight.

**Comment by kwizzn** [comment]
Sure!

```truffle-config.js```
```javascript
module.exports = {
  subscribers: {
    testnetOverride: require('./testnet-override.js'),
  },
  // ...
};
```

```testnet-override.js```
```javascript
const { readFileSync, writeFileSync } = require('fs');
module.exports = {
    handlers: {
        "compile:start": [
            function () {
                const source = readFileSync('some-contract.sol', 'utf8');
                const target = origin
                    .replace(/contract ClassName/g, 'contract ClassNameTestnet')
                    // more replacements here
                writeFileSync('some-new-contract.sol', target);
            }
        ],
    },
};
```

**Comment by cds-amal** [comment]
Hey @kwizzn, please be aware that the event system is for notification purposes and not to be used as lifecycle hooks.

**Comment by gnidan** [comment]
> Hey @kwizzn, please be aware that the event system is for notification purposes and not to be used as lifecycle hooks.

Yeah, definitely keep this caveat in mind!

... but honestly I'm impressed with your ingenuity here, and I'm not sure there's a clean alternative way to do what you want. I guess you don't want to duplicate your contracts :)

**Comment by kwizzn** [comment]
Not in production anyways, you're right. :)
